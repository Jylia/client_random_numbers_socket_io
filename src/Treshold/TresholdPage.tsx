import React, { Component } from 'react'
import openSocket from 'socket.io-client';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import { Button, Segment, Form } from 'semantic-ui-react';

import { History } from 'history';
import { ISocketData } from '../shared/ChartOptionInterfaces';

interface ITresholdPageProps {
    history: History;
}

interface ITresholdPageState {
    tresholdNumber?: number;
    allowedNumbers: Array<number>;
    isCalcInProgress: boolean;
}

class TresholdPage extends Component<ITresholdPageProps, ITresholdPageState> {
    socket?: SocketIOClient.Socket;

    constructor(props: ITresholdPageProps) {
        super(props);

        this.state = {
            allowedNumbers: [],
            isCalcInProgress: false,
        };
    }

    componentWillUnmount() {
        if (this.socket) this.socket.disconnect();
    }

    setNumber = (e: React.FormEvent<HTMLInputElement>) => {
        this.setState({ tresholdNumber: Number(e.currentTarget.value) });
    }

    notify = (value: number) => toast(`Number: ${value}`, {
        position: toast.POSITION.BOTTOM_LEFT,
        autoClose: 15000
    });

    checkRandomNumber = () => {
        this.setState({ isCalcInProgress: true });
        this.socket = openSocket('http://localhost:3001');
        this.socket.on('data', (data: ISocketData) => {
            if (this.state.tresholdNumber && data.value > this.state.tresholdNumber) {
                this.notify(data.value);
            }
        });
    }
    
    onCalcStopButtonClick = () => {
        this.setState({ isCalcInProgress: false });
        if (this.socket) this.socket.disconnect();
    }

    render () {
        return (
            <React.Fragment>
                <Segment inverted>
                    <Form inverted>
                        <Form.Field>
                            <label>Alert threshold</label>
                            <input
                                placeholder='Please, enter a number'
                                onChange={this.setNumber}
                            />
                        </Form.Field>

                        <Button
                            type='submit'
                            color={this.state.isCalcInProgress ? 'red' : 'teal'}
                            onClick={this.state.isCalcInProgress ? this.onCalcStopButtonClick : this.checkRandomNumber}
                        >
                        {this.state.isCalcInProgress ? 'Stop Calculation' : 'Check random number'}
                        </Button>
                    </Form>
                    <ToastContainer />
                </Segment>
            </React.Fragment>
        );
    }
}

export default TresholdPage