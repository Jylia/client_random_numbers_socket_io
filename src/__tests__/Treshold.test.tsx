import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import TresholdPage from '../Treshold/TresholdPage';
import { history } from '../History';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<TresholdPage history={history} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders correctly', () => {
  const tree = renderer
      .create(<TresholdPage history={history} />)
      .toJSON();
  expect(tree).toMatchSnapshot();
});
