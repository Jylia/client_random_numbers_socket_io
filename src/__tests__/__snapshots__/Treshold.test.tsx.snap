// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="ui inverted segment"
>
  <form
    className="ui inverted form"
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <label>
        Alert threshold
      </label>
      <input
        onChange={[Function]}
        placeholder="Please, enter a number"
      />
    </div>
    <button
      className="ui teal button"
      onClick={[Function]}
      type="submit"
    >
      Check random number
    </button>
  </form>
  <div
    className="Toastify"
  />
</div>
`;
